pkg/deploy/generator/resources.go

```diff
@@ -1089,6 +1078,116 @@ EOF
 
 chcon -R system_u:object_r:var_log_t:s0 /var/opt/microsoft/linuxmonagent
 
+cat >/usr/lib/systemd/system/download-mdm-credentials.service <<'EOF'
+[Unit]
+Description=Periodic mdm credentials refresh
+
+[Service]
+Type=oneshot
+ExecStart=/usr/local/bin/download-mdm-credentials.sh
+EOF
+
+cat >/usr/lib/systemd/system/download-mdm-credentials.timer <<'EOF'
+[Unit]
+Description=Periodic mdm credentials refresh
+
+[Timer]
+OnBootSec=0min
+OnCalendar=0/12:00:00
+AccuracySec=5s
+
+[Install]
+WantedBy=timers.target
+EOF
+
+cat >/usr/local/bin/download-mdm-credentials.sh <<EOF
+#!/bin/bash
+set -eu
+
+NEW_CERT_FILE="/etc/temp-mdm.pem"
+CURRENT_CERT_FILE="/etc/mdm.pem"
+
+az login -i
+az account set -s "$SUBSCRIPTIONID"
+
+rm -f \$NEW_CERT_FILE
+
+for attempt in {1..5}; do
+  az keyvault certificate download --file \$NEW_CERT_FILE --id "https://${KEYVAULTPREFIX}.vault.azure.net/certificates/test-cert" && break
+  if [[ \$attempt -lt 5 ]]; then sleep 10; else exit 1; fi
+done
+az logout
+
+if [ -f \$NEW_CERT_FILE ]; then
+  chmod 0600 \$NEW_CERT_FILE
+  sed -i -ne '1,/END CERTIFICATE/ p' \$NEW_CERT_FILE
+  mv \$NEW_CERT_FILE \$CURRENT_CERT_FILE
+else
+  echo Failed to refresh certificate && exit 1
+fi
+EOF
+chmod u+x /usr/local/bin/download-mdm-credentials.sh
+
+cat >/usr/lib/systemd/system/download-mdsd-credentials.service <<EOF
+[Unit]
+Description=Periodic mdsd certificate refresh
+
+[Service]
+Type=oneshot
+ExecStart=/usr/local/bin/download-mdsd-credentials.sh
+EOF
+
+cat >/usr/lib/systemd/system/download-mdsd-credentials.timer <<EOF
+[Unit]
+Description=Periodic mdsd certificate refresh
+
+[Timer]
+OnBootSec=0min
+OnCalendar=0/12:00:00
+AccuracySec=5s
+
+[Install]
+WantedBy=timers.target
+EOF
+
+cat >/usr/local/bin/download-mdsd-credentials.sh <<EOF
+#!/bin/bash
+set -eu
+
+NEW_CERT_FILE="/etc/temp-mdsd.pem"
+CURRENT_CERT_FILE="/etc/mdsd.pem"
+
+az login -i
+az account set -s "$SUBSCRIPTIONID"
+
+rm -f \$NEW_CERT_FILE
+
+for attempt in {1..5}; do
+  az keyvault certificate download --file \$NEW_CERT_FILE --id "https://${KEYVAULTPREFIX}.vault.azure.net/certificates/test-cert" && break
+  if [[ \$attempt -lt 5 ]]; then sleep 10; else exit 1; fi
+done
+az logout
+
+if [ -f \$NEW_CERT_FILE ]; then
+  chmod 0600 \$NEW_CERT_FILE
+  sed -i -ne '1,/END CERTIFICATE/ p' \$NEW_CERT_FILE
+  mv \$NEW_CERT_FILE \$CURRENT_CERT_FILE
+else
+  echo Failed to refresh certificate && exit 1
+fi
+EOF
+chmod u+x /usr/local/bin/download-mdsd-credentials.sh
+
+systemctl enable --now download-mdsd-credentials.timer
+systemctl enable --now download-mdm-credentials.timer
+
+/usr/local/bin/download-mdsd-credentials.sh
+
+/usr/local/bin/download-mdm-credentials.sh
+if [ $? != 0 ]; then
+  touch /etc/mdm.pem
+fi
```

Comment:
This is far from great. However the error case I am trying to work around by doing this will leave us at the same place as before. I think to fix the problem that we have currently is a meaty task that is better to handle outside of scope of my PR. 

If the mdm certificate was not downloaded for some reason and the file `/etc/mdm.pem` does not exist, then this line `-v /etc/mdm.pem:/etc/mdm.pem` in mdm.service definition will cause `/etc/mdm.pem` to be a directory rather than a file.

In the original code if mdm certificate was not downloaded properly, the mdm service will start, but will not be able to send data. This error handling case will cause the same behavior in case when certs are downloaded on a timer.

Reply:
this script is running -e: this can't happen.

Reply:
-v /etc/mdm.pem:/etc/mdm.pem: if /etcd/mdm.pem changes (specifically this case where a new file is mv'ed over the old file, which I like because it's atomic) does the container see the new file?

Reply:
the container does see the new file.
wrt `this script is running -e: this can't happen.` I have tested this and it is necessary. The variables that hold the new/curr path are escaped in the context of the "writer" script. and in the context of the download script itself - they are initialized with const values, it's just happens that one of them is a directory instead of a file. Then in this case the new cert is moved inside that current directory instead of the overwrite. (i.e. it will be `/etc/mdm.pem/temp-mdm.pem`

Reply:
> If the mdm certificate was not downloaded for some reason and the file /etc/mdm.pem does not exist,

If this happens, shouldn't we just let mdm fail (and be alerted via IcM), instead of trying to bury the errors?

> In the original code if mdm certificate was not downloaded properly, the mdm service will start, but will not be able to send data.

Sounds like mdm will be useless without a valid cert anyway, right? So I think we should just let it fail loudly.

Reply:
@ihcsim yes mdm will be useless without a certificate anyway, as it is right now. As I mentioned in the beginning of this thread, I am at the moment not aware how this failure is handled and I want t handle this as a separate task if possible. It might lead me into another rabbit hole and I really want to start moving this task ahead. Right now this PR will not change this behavior in the case when mdm certificate is not valid.

Reply:
re: `this script is running -e: this can't happen.` - yes, I wasn't right, indeed it can't happen. I should remove this "if" statement, mdm will still be broken but in another way

Reply:
if downloading the credentials first time around fails, the whole script should exit 1 (via -e) and that will cause the RP upgrade process to fail, which is fine for now.